README.md
# bettong

![bettong](https://upload.wikimedia.org/wikipedia/commons/thumb/3/3d/Bettongia_gaimardi.jpg/512px-Bettongia_gaimardi.jpg)

WIP: Bettong is a JavaScript Node.js web crawler based in Puppeteer. Based on the provided base URL, Bettong crawls pages on the same origin saves screenshots and HTML content.

## Requirements

Bettong uses [async/await](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function) which is only available in Node.js 8.x.x or higher.

## Options

| argument           | type               | required | default                                                                                                                          | description                                                                                                                                                                                 |
|--------------------|--------------------|----------|----------------------------------------------------------------------------------------------------------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| baseUrl            | string             | true     | null                                                                                                                             | URL to start crawling from. Bettong will only crawl pages that are on the same origin.                                                                                                      |
| outputPath         | string             | false    | "dist"                                                                                                                           | Relative path screenshots and/or html content will be saved to. Currently screenshots will be saved to provided-relative-path/screenshots and HTML content will be saved to provided-relative-path/html.                                                                                           |
| options            | object             | false    | {}                                                                                                                               | Bettong options.                                                                                                                                                                            |
| options.screenshot | boolean            | false    | true                                                                                                                             | Whether Bettong should save screenshot for each viewport.                                                                                                                                   |
| options.html       | boolean            | false    | true                                                                                                                             | Whether Bettong should save HTML content for each page.                                                                                                                                     |
| options.viewports  | puppeteer.Viewport | false    | [   { width: 540, height: 480, },   { width: 720, height: 480 },   { width: 960, height: 480 },   { width: 1140, height: 480 } ] | Array of viewports used to take screenshots. Only used if `options.screenshot` is set to `true`. Please puppeteer docs for more information on available properties for Viewport interface. |

## Usage

### Node.js

Install bettong

```sh
npm install --save bettong
```

```js
const Bettong = require('bettong');

const bettong = new Bettong('https://foo.bar');
await bettong.exec();
```

### CLI

```sh
Usage: bettong exec [options] <base-url>

Execute crawling starting from the required base url <base-url>

Options:
  -o, --output-path <path>         relative output path (default: "dist")
  -e, --exclude-pattern <pattern>  RegExp page exclude pattern (default: "")
  -s, --screenshot <screenshot>    save screenshots (default: true)
  -h, --html <screenshot>          save html content (default: true)
  -v, --viewport <viewport>        viewport for screenshots, e.g. '{"width":128,"height":128}'
  -h, --help                       output usage information
```

Install bettong globally

```sh
npm install -g bettong
```

#### Samples

Start crawling at https://foo.bar and exclude crawling pages that contain `'baz'` in the url.

```sh
bettong exec https://foo.bar -e '.*baz.*'
```

## Output

Bettong will save screenshots to `provided-relative-path/screenshots` and HTML content to `provided-relative-path/html`. By default this would be `dist/screenshots` and `dist/html`. A screenshot will be saved as a full page screenshot in PNG format for each viewport provided in `options.viewports`.


CHANGELOG.md
## [0.2.1](https://github.com/alexstaroselsky/bettong/compare/v0.2.0...v0.2.1) (2019-12-29)


### Bug Fixes

* 🐛 Added missing shebang line ([3082184](https://github.com/alexstaroselsky/bettong/commit/3082184b9d7b299caacd52b3d84a7a776906ac97))

# [0.2.0](https://github.com/alexstaroselsky/bettong/compare/v0.1.0...v0.2.0) (2019-12-29)


### Features

* 🎸 Added CLI functionality ([c40838d](https://github.com/alexstaroselsky/bettong/commit/c40838dbd58214b3f185966d7b80df9fe900b7b7))

# [0.1.0](https://github.com/alexstaroselsky/bettong/compare/v0.0.5...v0.1.0) (2019-12-25)


### Features

* 🎸 Added exclude pattern option ([8db9dd6](https://github.com/alexstaroselsky/bettong/commit/8db9dd6a50e3e6bd44f2bc6362e03f59cd7e5fd9))

## [0.0.5](https://github.com/alexstaroselsky/bettong/compare/v0.0.4...v0.0.5) (2019-12-25)


### Bug Fixes

* 🐛 Added logic to exclude pages based ([21d2a0a](https://github.com/alexstaroselsky/bettong/commit/21d2a0a9753a7fd5e6ea4d3c68ae3ccc41284baf))

## [0.0.4](https://github.com/alexstaroselsky/bettong/compare/v0.0.3...v0.0.4) (2019-12-25)


### Bug Fixes

* 🐛 Updated .releaserc to ensure npm plugin included ([3ab23f5](https://github.com/alexstaroselsky/bettong/commit/3ab23f54caa2d181ebe82a7e5af8353c27278342))

## [0.0.3](https://github.com/alexstaroselsky/bettong/compare/v0.0.2...v0.0.3) (2019-12-25)


### Bug Fixes

* 🐛 Updated logic for saving screenshots and html ([23e4c55](https://github.com/alexstaroselsky/bettong/commit/23e4c5523135b8d7da1eef5135e3214922f22924))

## [0.0.2](https://github.com/alexstaroselsky/bettong/compare/v0.0.1...v0.0.2) (2019-12-25)


### Bug Fixes

* 🐛 Updated logic for directory creation ([980e2b7](https://github.com/alexstaroselsky/bettong/commit/980e2b7dc7e5b0dc7002edcdeb6cebefe20d2b66))


lib/bettong.test.js
const Bettong = require('./bettong');

describe('Bettong', () => {
  it('should create an instance of Bettong', () => {
    expect(new Bettong('foo')).toBeInstanceOf(Bettong);
  });

  it('should merge options with default options', () => {
    const options = {
      html: false,
      viewports: [
        { width: 1, height: 1 }
      ]
    };
    const bettong = new Bettong('foo', 'bar', options);
    const expected = {
      screenshot: true,
      html: false,
      viewports: [
        { width: 1, height: 1 }
      ]
    };

    expect(bettong.options).toEqual(expected);
  });

  it('should exclude pages already added to visited', () => {
    const bettong = new Bettong('foo');
    bettong.visited.push('foo.bar');

    expect(bettong._isValidPage('foo.bar')).toEqual(false);
  });

  it('should exclude pages already added to queue', () => {
    const bettong = new Bettong('foo');
    bettong.queue.push('foo.bar');

    expect(bettong._isValidPage('foo.bar')).toEqual(false);
  });

  it('should exclude pages that match exclude pattern', () => {
    const bettong = new Bettong('foo', 'bar', { excludePattern: 'bar' });

    expect(bettong._isValidPage('foo.bar')).toEqual(false);
  });
});


lib/bettong.js
const _ = require('lodash');
const fs = require('fs');
const path = require('path');
const puppeteer = require('puppeteer');
const uuidv4 = require('uuid/v4');

/**
 * puppeteer.Viewport
 * @see {@link https://github.com/puppeteer/puppeteer}
 *
 * @typedef {object} Viewport
 * @property {number} width The page width in pixels.
 * @property {number} height The page height in pixels.
 * @property {number} [deviceScaleFactor=1] Specify device scale factor (can be thought of as dpr).
 * @property {boolean} [isMobile=false] Specify device scale factor (can be thought of as dpr).
 * @property {boolean} [hasTouch=false] Specifies if viewport supports touch events.
 * @property {boolean} [isLandscape=false] Specifies if viewport is in landscape mode.
 */

const defaultOptions = {
  screenshot: true,
  html: true,
  viewports: [
    { width: 540, height: 480, },
    { width: 720, height: 480 },
    { width: 960, height: 480 },
    { width: 1140, height: 480 },
  ]
};

/**
 * Bettong
 */
class Bettong {
  /**
   *
   * @param {string} baseUrl Base URL to start crawling from
   * @param {string} [outputPath="dist"] Relative path for screenshot and content output
   * @param {object} [options={}] Bettong options
   * @param {boolean} [options.screenshot=true] Whether screenshots should be taken and saved
   * @param {boolean} [options.html=true] Whether HTML content should be saved
   * @param {string} [options.excludePattern] RegExp expression to exclude pages from crawling
   * @param {Viewport[]} [options.viewports]
   */
  constructor(baseUrl, outputPath = 'dist', options = {}) {
    // TODO - Validate baseUrl
    this.baseUrl = baseUrl;
    // TODO - Validate outputPath
    this.outputPath = outputPath;
    // TODO - Validate options
    // TODO - Provide option for link selector
    this.options = _.assign({}, defaultOptions, this._sanitizeOptions(options));

    this.visited = [];
    this.queue = [];
  }

  /**
   * @access private
   *
   * Crawl page
   * @param {object} page
   */
  async _crawl(page) {
    const url = page.url();
    console.info(`Visiting: ${url}`);
    this.visited.push(url);

    // TODO - Use more human friendly value such as page url or title
    const uuid = uuidv4();

    if (this.options.html) {
      // Save page HTML
      const html = await page.content();
      fs.writeFileSync(`${this.relativePath}/html/${uuid}.html`, html);
    }

    if (this.options.screenshot) {
      // Save screenshot for every viewport
      for (const viewport of this.options.viewports) {
        console.info(`Taking full page screenshot of ${url} at size ${viewport.width}px by ${viewport.height}px`)
        await page.setViewport({ ...viewport });
        await page.screenshot({
          path: `${this.relativePath}/screenshots/${uuid}_${viewport.width}x${viewport.height}.png`,
          type: 'png',
          fullPage: true
        });
      }
    }

    const anchors = await page.$$('a');
    for (const anchor of anchors) {
      const hrefHandle = await anchor.getProperty('href');
      let href = await hrefHandle.jsonValue();
      // TODO - Provide option for keeping query params
      const idx = href.indexOf('#');
      if (idx > -1) {
        href = href.substring(0, idx);
      }

      if (this._isValidPage(href)) {
        this.queue.push(href);
      }
    }

    const nextUrl = this.queue.shift();
    if (!nextUrl) return;
    const nextPage = await this.context.newPage();
    await nextPage.goto(nextUrl, { waitUntil: 'load' });

    return this._crawl(nextPage);
  }

  /**
   * @access private
   *
   * Check if page should be crawled. Checks the following
   * - Does href include base url?
   * - Has href not been added to array of visisted pages?
   * - Has href not been added to array of queued pages?
   * - Does href not match exclude RegExp pattern?
   *
   * @param {string} href
   */
  _isValidPage(href) {
    let isValid = false;

    isValid = href.includes(this.baseUrl) &&
      !_.includes(this.visited, href) &&
      !_.includes(this.queue, href);

    if (this.options.excludePattern) {
      const regex = new RegExp(this.options.excludePattern, 'g');
      isValid = isValid && !regex.test(href);
    }

    if (!isValid) {
      console.info(`Excluding crawling page ${href}`);
    }

    return isValid;
  }

  /**
   * @access private
   *
   * Remove undefined values from options object
   *
   * @param {object} options
   * @returns {object}
   */
  _sanitizeOptions(options) {
    return _.transform(options, function (result, value, key) {
      if (!_.isNil(value)) result[key] = value;
    }, {});
  }

  /**
   * Execute crawling
   *
   * @access public
   */
  async exec() {
    this.relativePath = path.resolve(this.outputPath);
    if (!fs.existsSync(this.relativePath)) {
      if (this.options.screenshot) {
        // Create directory for screenshot output
        console.info(`Creating sceenshot output directory at ${this.relativePath}/screenshots`);
        fs.mkdirSync(`${this.relativePath}/screenshots`, { recursive: true });
      }
      if (this.options.html) {
        // Create directory for html output
        console.info(`Creating html output directory at ${this.relativePath}/html`)
        fs.mkdirSync(`${this.relativePath}/html`, { recursive: true });
      }
    }

    const browser = await puppeteer.launch();
    this.context = await browser.createIncognitoBrowserContext();
    const page = await this.context.newPage();
    await page.goto(this.baseUrl, { waitUntil: 'load' });

    if (this._isValidPage(page.url())) {
      // TODO - Prevent redirects
      // https://github.com/puppeteer/puppeteer/issues/1132#issuecomment-393724933
      await this._crawl(page);
    }

    await this.context.close();
  }
}

module.exports = Bettong;
